### PR TITLE
Snowflake query tag

### LINKED ISSUE #1030: Add query_tag to the model hierarchy
### Feature description

After starting to analyze our query history in some depth, I had the realization that it would be useful to be able to specify QUERY_TAG (snowflake concept) anywhere in the hierarchy, like the `materialization` configuration.

This way, you can query snowflake query_history by the query tag. Maybe you set one query tag for your whole dbt project. Maybe you set it per model folder. Maybe you set it on specific models.

https://docs.snowflake.net/manuals/sql-reference/parameters.html#query-tag
https://stackoverflow.com/questions/50965558/snowflake-cli-snowsql-query-tagging

QUERY_TAG is a session parameter, so the solution might be to set it in snowflake's table and view materializations, if the query tag is set?

### Who will this benefit?

Anyone using snowflake who would like more granular information in her query history.

### PR DESCRIPTION
**Pull request by DrMcTaco**
resolves #1030

### Description
Add the ability to set snowflake query tags upon creation of a connection or at the start of a materialization. For query tags used during materialization the session parameter that controls query tags will be returned to whatever value it was before the materialization began

Not sure what/how to test for this addition.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: plugins/snowflake/dbt/include/snowflake/macros/adapters.sql [hunk 1]
```diff
@@ -160,3 +160,33 @@
...
+    {% set original_query_tag = get_current_query_tag() %}
+    {{ log("Setting query_tag to '" ~ new_query_tag ~ "'. Will reset to '" ~ original_query_tag ~ "' after materialization.") }}
+    {% do run_query("alter session set query_tag = '{}'".format(new_query_tag)) %}
+    {{ return(original_query_tag)}}
+  {% endif %}
+  {{ return(none)}}
+{% endmacro %}
+
+{% macro unset_query_tag(original_query_tag) -%}
+  {% set new_query_tag = config.get('query_tag', none) %}
```
Comment (beckjake) [inline plugins/snowflake/dbt/include/snowflake/macros/adapters.sql:182]:
```suggestion
  {% set new_query_tag = config.get('query_tag') %}
```

There's no reasonable way for you to know this, but `config.get` doesn't do what you think it does! The second argument isn't a default value, but a "validator". See #2441.

**Comment by DrMcTaco** [comment]
> It would also be great if you could add tests, though if tests aren't really reasonable here I understand that. Does snowflake have an API or something we could query that shows this stuff?

Snowflakes query tags are session level parameters. They can be viewed by running the sql `SHOW PARAMETERS [ LIKE '<pattern>' ][ IN { SESSION | ACCOUNT | ... }]`. The query_tag parameter does not appear to be accessible by introspecting the snowflake connection object and I am not aware of any way to find session parameters for arbitrary open sessions. So any kind of testing would require running the SQL in the [`get_current_query_tag` macro](https://github.com/fishtown-analytics/dbt/pull/2555/files#diff-b197e42113386f8060c73c4dd9cf4054R165-R167) from inside the session at appropriate times or looking at the snowflake [query_history view](https://docs.snowflake.com/en/sql-reference/account-usage/query_history.html), which has some obnoxious lag. I think given the mentioned complications testing for this feature might be too difficult.

> One thing I would like to bring up: On errors, it seems to me that this will result in the query tag getting "stuck", because the unset_query_tag never runs, right? Is that a big problem? If it is, I think we could add a dbt feature to make cleanup more feasible (some sort of adapter-level pre/post model hooks, or a finally equivalent in jinja, or something similar).

In my view query_tags are primarily a tacking and observability tool, not a functional part of the transformation process. To my knowledge they cannot impact the running of SQL against snowflake. So "getting stuck" with an incorrect query_tag will not stop dbt from running properly. 

Also I am not familiar enough with how dbt handles SQL failures in the middle of runs with multiple materializations. Will connections/sessions be reused for subsequent models if one model fails?

**Comment by jtcohen6** [comment]
The stakes here feel pretty low. I haven't used query tags in Snowflake, but I could imagine them being most compelling as a mechanism for longitudinal analysis of models across production runs where builds are almost always successful, e.g. to measure performance timing. If I needed to debug a run with failed models from within Snowflake's query history, I'd more likely look to the `query-comment` or raw SQL as a starting point.

That said, I agree that we should have a construct for end-of-materialization cleanup steps that aren't precluded by model SQL errors.

**Comment by beckjake** [comment]
Cool, let's just add this as-is then. We can add support for proper cleanups at a later date. I'm going to test it out locally manually before I kick off tests - I will try to get to this today, but it may be next week.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,6 @@
+## Next
+- Added support for Snowflake query tags at the connection and model level ([#1030](https://github.com/fishtown-analytics/dbt/issues/1030))
```
Comment (beckjake) [inline CHANGELOG.md:2]: Can you please move this into the `dbt 0.18.0` section, put it under a `### Features` heading, and make it link to both the issue and this PR?

**Comment by beckjake** [comment]
This looks ready to go to me! I've merged the changelog updates and once the tests pass I'll merge this in for 0.18.0. Thanks for your contribution, @DrMcTaco!

**Pull request merged**